nemo/core/neural_factory.py

```diff
@@ -28,18 +28,210 @@
+        args:
+            action (Actions): The Actions object this state is associated with.
+        """
+        tensor_naming_registery = AppState().tensor_names
+        self.tensor_dict = {}.fromkeys(tensor_naming_registery.unique_names, None)
+        self._action = action
+
+    def tensor_list(self):
+        """Returns a list the unique names of all tensors.
+        """
+        return self.tensor_dict.keys()
+
+    def clear_dict(self):
+        """Clears the dictionary by setting all values to None. Used in-between training batches to clear it's state.
+        """
+        for name in self.tensor_dict:
+            self.tensor_dict[name] = None
+
+    def set_tensor(self, tensor: NmTensor, value: 'torch.Tensor'):
+        """Sets the value of tensor
+
+        args:
+            tensor (NmTensor)
+            value (torch.Tensor)
+        """
+        self.tensor_dict[tensor.unique_name] = value
+
+    def check_tensor_cached(self, unique_name: str):
+        """Checks to see the tensor value has been computed in the current step yet.
+
+        args:
+            unique_name (str): The NmTensor.unique_name that we want to check for.
+
+        returns:
+            (bool) whether the tensor with unique_name has been computed yet.
+        """
+        if self.tensor_dict[unique_name] is None:
+            return False
+        return True
+
+    def get_tensor(self, name: Union[str, NmTensor], compute: bool = True):
+        """Returns the value associated with a tensor. And optionally, computes the value of the tensor if not already
+        set.
+
+        args:
+            name (str, NmTensor): The user-defined name for a tensor or the NmTensor itself.
+            compute (bool): If True and the tensor has not already been computed, there will be an attempt to create a
+                call DAG and then do a forward pass on this call DAG to compute the tensor. If False, it will return
+                None if the tensor has not been computed yet.
+                Defaults to True.
+
+        returns:
+            (torch.tensor or None) representing the computed value of the requested name. Returns None if compute is
+            False and the tensor has not been computed yet.
+        """
+        if isinstance(name, NmTensor):
+            unique_name = name.unique_name
+        else:
+            unique_name = AppState().tensor_names[name]
+        tensor_value = self.tensor_dict[unique_name]
+        if tensor_value is None and compute:
+            nmtensor = AppState().tensor_names._nmtensor_uniname_dict[unique_name]
+            callchain = topological_sort_from_leaves([nmtensor], cached_training_state=self)
+            callchain.insert(0, ())
+            self._action.nm_graph_forward_pass(callchain, self.tensor_dict)
+            tensor_value = self.tensor_dict[unique_name]
+        return tensor_value
+
+
+def topological_sort_from_leaves(leaf_nmtensors: List[NmTensor], cached_training_state: TrainingState = None):
```

Comment:
Ok, really I don't get it. Why this code, that is totally unrelated with Neural(Module)Factory, landed in here??

Reply:
Moved topological_sort_from_leaves back into backends\pytorch\actions.py